pkg/eventingtls/eventingtls.go

```diff
@@ -0,0 +1,200 @@
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	// TLSKey is the key in the TLS secret for the private key of TLS servers
+	TLSKey = "tls.key"
+	// TLSCrt is the key in the TLS secret for the public key of TLS servers
+	TLSCrt = "tls.crt"
+	// DefaultMinTLSVersion is the default minimum TLS version for servers and clients.
+	DefaultMinTLSVersion = tls.VersionTLS12
+)
+
+type ClientConfig struct {
+	// CACerts are Certification Authority (CA) certificates in PEM format
+	// according to https://www.rfc-editor.org/rfc/rfc7468.
+	CACerts *string
+}
+
+type ServerConfig struct {
+	// GetCertificate returns a Certificate based on the given
+	// ClientHelloInfo. It will only be called if the client supplies SNI
+	// information or if Certificates is empty.
+	//
+	// If GetCertificate is nil or returns nil, then the certificate is
+	// retrieved from NameToCertificate. If NameToCertificate is nil, the
+	// best element of Certificates will be used.
+	GetCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+}
+
+// GetCertificate returns a Certificate based on the given
+// ClientHelloInfo. It will only be called if the client supplies SNI
+// information or if Certificates is empty.
+//
+// If GetCertificate is nil or returns nil, then the certificate is
+// retrieved from NameToCertificate. If NameToCertificate is nil, the
+// best element of Certificates will be used.
+type GetCertificate func(*tls.ClientHelloInfo) (*tls.Certificate, error)
+
+// GetCertificateFromSecret returns a GetCertificate function that will automatically return
+// the latest certificate that is present in the provided secret.
+//
+// The secret is expected to have at least 2 keys in data: see TLSKey and TLSCrt constants for
+// knowing the key names.
+func GetCertificateFromSecret(ctx context.Context, informer coreinformersv1.SecretInformer, kube kubernetes.Interface, secret types.NamespacedName) GetCertificate {
+
+	certHolder := atomic.Value{}
+
+	logger := logging.FromContext(ctx).Desugar().
+		With(zap.String("tls.secret", secret.String()))
+
+	store := func(obj interface{}) {
+		s, ok := obj.(*corev1.Secret)
+		if !ok {
+			return
+		}
+		crt, crtOk := s.Data[TLSCrt]
+		key, keyOk := s.Data[TLSKey]
+		if !crtOk || !keyOk {
+			logger.Debug("Missing " + TLSCrt + " or " + TLSKey + " in the secret.data")
+			return
+		}
+
+		logger.Debug("Loading key pair")
+
+		certificate, err := tls.X509KeyPair(crt, key)
+		if err != nil {
+			return
+		}
+
+		logger.Debug("certificate stored")
+		certHolder.Store(&certificate)
+	}
+
+	informer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: controller.FilterWithNameAndNamespace(secret.Namespace, secret.Name),
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: store,
+			UpdateFunc: func(_, newObj interface{}) {
+				store(newObj)
+			},
+			DeleteFunc: nil,
```

Comment:
Should we clear the store on a delete? I can't think of a valid reason for a user to delete the Secret but the intent is likely to stop the usage of the cert when that happens.

Reply:
I was thinking about that, however, if the secret is deleted by mistake that would cause an outage or maybe the cert management is migrated to a different tool (like Cert-Manager first and Vault after), there might be a need to delete and recreate the secret without downtime and that is hard if we stop serving the certs.

Reply:
That makes sense to me